feat(queries): 'messages' query type for structured conversations and multi-model input

adriander opened the pull request:
# feat: Add flexible input handling with RawExtension for Query resources

## Summary
Refactors the Query system to support flexible input types using `runtime.RawExtension`, enabling both simple string inputs and structured message arrays across the ark-controller, ark-API, and ark-dashboard.

## Key Changes

### 🔧 Backend (ark-Controller)
- **Enhanced Query CRD**: Changed `Input` field from `string` to `runtime.RawExtension` with type discrimination (`user` vs `messages`)
- **Helper Methods**: Implemented `GetInputString()`, `GetInputMessages()`, `SetInputString()`, `SetInputMessages()` for type-safe access

### 🐍 API Layer (ark-API)
- **Model Updates**: Added `InputType` enum and `Message` model for structured conversation data
- **Request/Response Models**: Updated `QueryCreateRequest`, `QueryDetailResponse` to support `Union[str, List[Message]]` input
- **Smart Input Handling**: API automatically determines input type and properly serializes for RawExtension

### ⚛️ Frontend (ark-Dashboard)
- **Dual Input Modes**: Chat interface supports both string concatenation and message array approaches
- **UI Toggle**: Users can switch between "String" (S) and "Messages" (M) input modes 
- **New Chat Service**: `submitFlexibleChatQuery()` method handles both input types seamlessly
- **Type Safety**: TypeScript support for `Message[]` and `InputType`

## Benefits
- **Multi-turn Conversations**: Support for structured conversation history with proper role attribution
- **Backward Compatibility**: Existing string-based queries continue to work unchanged
- **Type Safety**: End-to-end type safety from ARK-Dashboard → ARK-API → ARK-Controller
- **Developer Choice**: Developers can choose the most appropriate input method for their use case

## Breaking Changes
None - all changes are backward compatible. Existing string-based queries default to `type: "user"` and work unchanged.

## Example queries
### Query payload with input as string ARK-Dashboard -> ARK-API
```
{
  "name": "chat-query-740e1bf4-45a7-4f28-bbd2-0daa74f5df7e",
  "type": "user",
  "input": "User: m1. calculate 1+1.\n\nAgent: To calculate 1 + 1, the sum is equal to 2.\n\nUser: m2. calculate 2+2.\n\nAgent: To calculate 2 + 2, the sum is equal to 4.\n\nUser: m3. calculate the sum of previous sums.",
  "targets": [{ "type": "agent", "name": "sample-agent" }],
  "sessionId": "session-1758715375281"
}
```

### Query payload with input as []Messages ARK-Dashboard -> ARK-API
```
{
  "name": "chat-query-49edf480-17b8-459c-aec0-b8413f1723a9",
  "type": "messages",
  "input": [
    { "role": "user", "content": "m1. calculate 1+1." },
    { "role": "assistant", "content": "2" },
    { "role": "user", "content": "m2. calculate 2+2." },
    { "role": "assistant", "content": "4" },
    { "role": "user", "content": "m3. calculate the sum of previous sums." }
  ],
  "targets": [{ "type": "agent", "name": "sample-agent" }],
  "sessionId": "session-1758715485792"
}

```

adriander committed:
feat: enhance query models with InputType and Message support

adriander committed:
Merge remote-tracking branch 'origin' into add-chat-msg-to-queries

adriander committed:
update ark api usage in dashboard

adriander committed:
add flexible input handling for chat queries and enhance message processing

adriander committed:
update comment for manifests target to clarify CRD generation

adriander committed:
Merge remote-tracking branch 'origin' into add-chat-msg-to-queries

adriander committed:
enhance query handling with flexible input types and message support

adriander committed:
implement GetQueryInputMessages function for flexible message handling

adriander committed:
rename to historyMessages

adriander committed:
refactor query execution to use GetQueryInputMessages for improved message handling

adriander committed:
refactor query execution to use GetQueryInputMessages for improved message handling

adriander committed:
refactor executeModel to use GetQueryInputMessages for improved message handling

adriander committed:
refactor executeTarget to use GetQueryInputMessages for improved telemetry

adriander committed:
visual fix

adriander committed:
Merge remote-tracking branch 'origin' into add-chat-msg-to-queries

adriander committed:
test: add unit tests for GetQueryInputMessages function

adriander committed:
fix linter

adriander committed:
fix: update query_to_detail_response to use metadata for name and namespace

adriander committed:
use rawextension for input

adriander committed:
feat: add input handling methods for QuerySpec to support string and message types

adriander committed:
enhance GetQueryInputMessages to handle user input and messages more effectively

adriander committed:
todo. add temporary workaround

adriander committed:
fix: improve error handling and telemetry in executeTarget function

adriander committed:
refactor: streamline input message handling in executeTarget and executeAgent functions

adriander committed:
refactor: update executeTarget and related methods to use responseMessages for improved clarity

adriander committed:
refactor: update input field to accept both string and array of messages for improved flexibility

adriander committed:
update EvaluationRequest input type to support array of messages

adriander committed:
update test cases to use SetInputString and SetInputMessages

adriander committed:
update QuerySpec to support schemaless input and improve test cases for input handling

adriander committed:
fix: update e2e test to use build-container command for image creation

adriander committed:
streamline input handling in query response

adriander committed:
update ark-sdk dependency version to 0.1.34 in pyproject.toml

adriander committed:
clean up whitespaces

adriander committed:
fix lint

adriander committed:
add: include 'Type' column in additionalPrinterColumns for queries CRD

dwmkerr commented on ark/api/v1alpha1/query_types.go:
```diff
@@ -3,9 +3,30 @@
+)
+
+const (
+	// QueryTypeUser represents a query with string input
```
which is converted into an openai completions message of type `{"role": "user"` might be too much but maybe useful, same for the comments below to be super explicit as I worry it could confuse people at first

adriander replied:
done
```
    // QueryTypeUser represents a query with string input that gets converted to a single message with role="user"
	QueryTypeUser = "user"
	// QueryTypeMessages represents a query with an array of Message objects, each with explicit role and content
	QueryTypeMessages = "messages"
	```

dwmkerr commented on ark/api/v1alpha1/query_types.go:
```diff
@@ -3,9 +3,30 @@
 )
 
+// Message represents a single message in a conversation
+type Message struct {
```
again should it potentially be an OpenAI message or something? but am not too picky on this one

adriander replied:
yes. that makes sense. Let's have it in a followup. right now there would be cycle imports and that's requires more refactoring to resolve

dwmkerr replied:
Just an FYI while i was testing locally I changed to deseralize into the openai type and worked ok with a full e2e test so have added that here while doing the full run through

dwmkerr commented on ark/api/v1alpha1/query_types.go:
```diff
@@ -32,9 +53,15 @@ type EvaluatorRef struct {
+	// +kubebuilder:validation:Optional
+	// +kubebuilder:validation:Enum=user;messages
+	// +kubebuilder:default=user
+	Type string `json:"type,omitempty"`
```
@havasik should we do the admission conversion to `user` by default so that we always have a value, consistent with what `model` will be for query (and `type` `ClusterIP` for services, if i remember correctly)

dwmkerr commented on ark/api/v1alpha1/query_types.go:
```diff
@@ -118,6 +145,73 @@ type QueryList struct {
+	}
+
+	var messages []Message
+	if err := json.Unmarshal(q.Input.Raw, &messages); err != nil {
```
what happens if the user uses other valid stuctures such as tool call messages? do we preserve them and send them on? the spec says we should forward all of the content [and this will be a *huge* help cause upcoming or model specific parameters are then supproted out of the box)

adriander replied:
Great point. currently other fields are silently ignored. Do we want to address this now? or in a follow up? What should the Message structure be?

dwmkerr commented on ark/internal/controller/query_controller.go:
```diff
@@ -477,40 +477,68 @@ func (r *QueryReconciler) finalize(ctx context.Context, query *arkv1alpha1.Query
+	var err error
+	metadata := map[string]string{"targetType": target.Type, "targetName": target.Name}
+
+	// Get input messages for processing and telemetry
```
is this actually for telemetry? or is it that if it fails we record the error to the temetry provider, cause the actual messages are logged later on as they were originally?

adriander replied:
It is for the error events. telemetry is not using this. `BaseEvent` requires `metadata`. It was in main before, so I kept it

dwmkerr commented on ark/internal/controller/query_controller.go:
```diff
@@ -552,33 +580,35 @@ func (r *QueryReconciler) executeAgent(ctx context.Context, query arkv1alpha1.Qu
-	}
-
-	userMessage := genai.NewUserMessage(resolvedInput)
+	// Execute agent with the last message as the current input and previous messages as context
```
this i think we'd need unit tests for and maybe even a chainsaw test [chainsaw can come later i expect, but memory is something i think we'll get issues with without robust tests)

adriander replied:
absolutely. That is for an architecture discussion on how to work with memory _and_ input messages. 

my thinking was, that on a timeline the memory is written before the new InputMessages are received. Therefore I decided to append the input messages after the memory message. 
Let me know if you have a different view here and I am happy to update the code.

dwmkerr commented on ark/internal/controller/query_controller.go:
```diff
@@ -591,34 +621,34 @@ func (r *QueryReconciler) executeTeam(ctx context.Context, query arkv1alpha1.Que
-	}
-
-	userMessage := genai.NewUserMessage(resolvedInput)
+	// Execute team with the last message as the current input and previous messages as context
```
why don't we extract these methods into `internal` `memory` or something as independent helper functions with their own unit tests

adriander replied:
helper function created

dwmkerr commented on ark/internal/controller/query_controller.go:
```diff
@@ -591,34 +621,34 @@ func (r *QueryReconciler) executeTeam(ctx context.Context, query arkv1alpha1.Que
-	// Save new messages to memory (user message + response messages)
-	newMessages := append([]genai.Message{userMessage}, responseMessages...)
+	// Save all new messages (input + response) to memory
+	newMessages := make([]genai.Message, 0, len(inputMessages)+len(responseMessages))
```
same

dwmkerr commented on ark/internal/controller/query_controller.go:
```diff
@@ -631,22 +661,15 @@ func (r *QueryReconciler) executeModel(ctx context.Context, query arkv1alpha1.Qu
-	// Append user message to conversation history
-	messages = append(messages, userMessage)
-	allMessages := messages
+	// Append all input messages to conversation history
```
same

dwmkerr commented on ark/internal/controller/query_controller.go:
```diff
@@ -678,16 +701,18 @@ func (r *QueryReconciler) executeModel(ctx context.Context, query arkv1alpha1.Qu
 }
 
-func (r *QueryReconciler) executeTool(ctx context.Context, query arkv1alpha1.Query, toolName string, impersonatedClient client.Client, tokenCollector *genai.TokenUsageCollector) ([]genai.Message, error) { //nolint:unparam
+func (r *QueryReconciler) executeTool(ctx context.Context, query arkv1alpha1.Query, inputMessages []genai.Message, toolName string, impersonatedClient client.Client, tokenCollector *genai.TokenUsageCollector) ([]genai.Message, error) { //nolint:unparam
```
should we ever do this or should content of type `messages` never go to tools, cause it's pretty much by definition wrong

dwmkerr commented on ark/internal/genai/evaluator.go:
```diff
@@ -80,7 +80,11 @@ func CallSingleEvaluator(ctx context.Context, k8sClient client.Client, query ark
 	}
 
 	// For backward compatibility with query evaluation - evaluation CRDs don't use this path
```
CC for @skokaina this looks like stale code should we kill it?

dwmkerr commented on ark/internal/genai/query_parameters.go:
```diff
@@ -126,3 +126,49 @@ func ResolveBodyTemplate(ctx context.Context, k8sClient client.Client, namespace
+
+		// Convert CRD messages to genai.Message format
+		var messages []Message
+		for _, msg := range crdMessages {
```
checking again i think we need to (potentially) just keep whatever comes in, role might later have `thinking` etc [i think some providers overload role]

adriander replied:
right. change it to throw an error

dwmkerr commented on ark/internal/genai/query_parameters_test.go:
```diff
@@ -0,0 +1,365 @@
+		}
+	})
+
+	b.Run("messages_type", func(b *testing.B) {
```
not quite relevant for here but i think a test to make sure additional `messages` params are preserved will be useful

dwmkerr commented on services/ark-dashboard/ark-dashboard/components/sections/queries-section.tsx:
```diff
@@ -85,6 +85,17 @@ export const QueriesSection = forwardRef<{ openAddEditor: () => void }, QueriesS
   };
 
+  // Helper function to convert input to displayable string
+  const getInputDisplayText = (input: string | { role: string; content: string; }[] | undefined): string => {
```
would it be better to show this as json?

dwmkerr commented on services/ark-dashboard/ark-dashboard/components/floating-chat.tsx:
```diff
@@ -139,6 +141,24 @@ export default function FloatingChat({ name, type, namespace, position, onClose
 
+  // New function to build message array instead of concatenated string
+  const buildChatMessages = (messages: ChatMessageData[], currentMsg: string): Message[] => {
+    const messageArray: Message[] = messages
```
out of curiosity, on the client side, we could actualy use OpenAI message types couldn't we? I do that in the `ark` CLI kind of to enforce that as a consumer and I can use openai types for messages - less important but worth considering

dwmkerr commented on services/ark-dashboard/ark-dashboard/lib/api/generated/types.ts:
```diff
@@ -2474,8 +2492,10 @@ export interface components {
+            type: components["schemas"]["InputType"] | null;
             /** Input */
-            input: string;
+            input: string | components["schemas"]["Message"][];
```
also, thinking about this, is it really openai messages we have here as that is what we're providing compatibility / extensibility for?

adriander replied:
right now the Message struct is very basic (role & content). But we can extend it

dwmkerr commented on services/ark-mcp/ark-mcp/pyproject.toml:
```diff
@@ -79,4 +79,4 @@ testpaths = [
 
 [tool.uv.sources]
-ark-sdk = { path = "../../../out/ark-sdk/py-sdk/dist/ark_sdk-0.1.31-py3-none-any.whl" }
+ark-sdk = { path = "../../../out/ark-sdk/py-sdk/dist/ark_sdk-0.1.34-py3-none-any.whl" }
```
No where else to put this in the file by file review - there's a deterministic query test that uses a mocked echo OpenAI endpoint so that you can actually deterministically test the final resolved query content to the LLM - should we just slightly improve that server and actually chainsaw test that parameter expansion and messages query input and messages with additional fields/structure input is sent properly?

adriander replied:
chainsaw tests are added

adriander committed:
fix: enhance comments for QueryTypeUser and QueryTypeMessages for clarity

adriander committed:
refactor: extract message preparation logic into separate helper functions

adriander committed:
fix: update GetQueryInputMessages to return error for unknown message roles

adriander committed:
lint

adriander committed:
add: create new query files for conversations

adriander committed:
add query example usage to docs

adriander committed:
docs: enhance query input schema with detailed examples and notes

adriander committed:
add: implement Chainsaw tests for query input types and validation

adriander committed:
Merge remote-tracking branch 'origin' into rawextension-input-query

adriander committed:
Merge remote-tracking branch 'origin' into rawextension-input-query

adriander committed:
fix: add missing print column for Response type in query_types.go

adriander committed:
refactor: remove deprecated evaluator functions for cleaner codebase

adriander committed:
refactor: remove buildEvaluationRequest function for cleaner code

adriander committed:
Merge remote-tracking branch 'origin' into rawextension-input-query

adriander committed:
in fark, use runtime.RawExtension for input handling

adriander committed:
Merge remote-tracking branch 'origin' into rawextension-input-query

adriander committed:
Merge remote-tracking branch 'origin' into rawextension-input-query

adriander committed:
fix: change Input type to runtime.RawExtension in Query struct

adriander committed:
fix: update test name to evaluation-direct-ragas for clarity

adriander committed:
Merge remote-tracking branch 'origin' into rawextension-input-query

adriander committed:
refactor: simplify input handling in createQuery and related functions

adriander committed:
Merge remote-tracking branch 'origin' into rawextension-input-query

dwmkerr committed:
Merge branch 'main' into rawextension-input-query

dwmkerr committed:
chore: clean up samples

dwmkerr committed:
fix: add streaming support to A2A execution engine

The A2A execution engine did not support streaming responses. Per the
streaming spec, non-streaming backends should send the complete response
as a single chunk at the end of execution.

Changes:
- Pass eventStream parameter through A2A execution path
- Send complete response as single chunk if streaming is enabled
- Use query ID as completion ID and agent/name format for model ID
- Follows same pattern as Bedrock non-streaming implementation

dwmkerr committed:
docs: restructure query and model reference documentation

Move query and model documentation to dedicated resource pages in
/reference/resources/ and retire sections from the CRDs overview page.

Changes:
- Add /reference/resources/query.mdx with complete query documentation
  - Input types (user/messages)
  - Targets (agent/team/model/tool)
  - Query parameter expansion with agent parameter references
  - Session management and examples
  - Status and phases
- Update /reference/resources/models.mdx
  - Add Model Properties section with example
  - Add Status and Health Checking section
  - Keep Agent Model Configuration as final section
- Remove Query and Models sections from /reference/crds.mdx
- Remove /reference/models.mdx (moved to /reference/resources/models.mdx)
- Update sidebar navigation in both _meta.js files

dwmkerr committed:
fix: sync dist CRDs with latest changes (#237)

## Summary
- Add baseUrl field for Bedrock models (from PR #124)
- Add queryParameterRef support in ValueSource fields (from PR #140)
- Remove evaluator-related fields from Query CRD
- Remove conditional wrapper around serviceaccounts impersonation RBAC
permission

Regenerated using kubebuilder, carefully staged only meaningful field
additions/removals (no whitespace changes).

dwmkerr committed:
feat: use OpenAI ChatCompletionMessageParamUnion for query messages (#238)

## Summary
- Remove CRD Message struct (only had role/content fields)
- Query messages now use `openai.ChatCompletionMessageParamUnion`
directly
- Supports full OpenAI message features: `name`, `tool_calls`,
`image_url`, multimodal content
- Agent responses include agent name in message metadata via
`processAssistantMessage`
- Add `NewAssistantMessageWithName` helper for team queries where
multiple agents contribute

dwmkerr committed:
docs: add multimodal message support to query reference

dwmkerr committed:
refactor: remove string mode chat and user: prefix stripping

- Remove inputMode toggle and string concatenation from floating chat
- Always use messages array for chat queries
- Remove cleanUserInput function that stripped 'user:' prefix
- Simplify chat input placeholder

dwmkerr committed:
fix: restore conditional serviceaccount impersonation in RBAC

dwmkerr committed:
refactor: remove unused NewAssistantMessageWithName function

The name is set directly in processAssistantMessage, this helper was never used.

dwmkerr committed:
refactor: simplify chat service to only use message arrays

- Remove submitFlexibleChatQuery (handled both string and Message[])
- Simplify submitChatQuery to only accept Message[]
- Always use type: messages for chat queries

dwmkerr committed:
feat: use OpenAI message types throughout and remove role prefix hack

- Replace custom Message model with ChatCompletionMessageParam
- OpenAI completions endpoint now creates type: messages queries
- Remove hacky role: prefix concatenation in completions API
- Pass messages array directly to queries (supports multimodal content)
- Update dashboard queries table to show message content without role prefix
- Update token counting to work with message arrays

dwmkerr committed:
fix: lint and test fixes for ark-cli, ark-api, and ark-dashboard

- Fix unused error variable in ark-cli waitForReady.ts
- Remove unused imports from ark-api queries.py
- Fix dashboard chat test to use message array instead of string
- Fix TypeScript type in dashboard queries-section (use unknown instead of any)
- Auto-fixed lint issues across ark-cli files

dwmkerr committed:
refactor: use OpenAI SDK types directly in dashboard

- Install openai package in dashboard
- Replace generated Message type with ChatCompletionMessageParam
- Add type assertion for input to handle multimodal content
- Maintains compatibility with API while using proper OpenAI types

dwmkerr committed:
refactor: use ChatCompletionMessageParam directly without Message alias

Remove unnecessary Message type alias in chat service and use OpenAI's
ChatCompletionMessageParam directly throughout the codebase.

dwmkerr committed:
refactor: use ChatCompletionMessageParam directly in chat UI

Replace ChatMessageData wrapper with ChatCompletionMessageParam to preserve
all OpenAI message fields (name, tool_calls, etc.) without mapping.

dwmkerr committed:
refactor: regenerate API types after removing Message schema

- Remove unused InputType export from chat service
- Regenerate dashboard types from updated ark-api openapi.json
- Use type assertion for query creation due to openapi-typescript limitations with union types

dwmkerr committed:
fix: make content optional in queries table input display

dwmkerr committed:
chore: update go.sum for openai-go dependency

dwmkerr committed:
chore: update fark go.mod for ark openai-go dependency
